src/coreclr/jit/flowgraph.cpp

```diff
@@ -484,256 +480,292 @@ PhaseStatus Compiler::fgExpandStaticInit()
-                //
+    fgMorphStmtBlockOps(block, stmt);
+    gtUpdateStmtSideEffects(stmt);
 
-                // Unlink block and prevBb
-                fgRemoveRefPred(block, prevBb);
+    //
+    // Update preds in all new blocks
+    //
 
-                // Block has two preds now: either isInitedBb or helperCallBb
-                fgAddRefPred(block, isInitedBb);
-                fgAddRefPred(block, helperCallBb);
+    // Unlink block and prevBb
+    fgRemoveRefPred(block, prevBb);
 
-                // prevBb always flow into isInitedBb
-                fgAddRefPred(isInitedBb, prevBb);
+    // Block has two preds now: either isInitedBb or helperCallBb
+    fgAddRefPred(block, isInitedBb);
+    fgAddRefPred(block, helperCallBb);
 
-                // Both fastPathBb and helperCallBb have a single common pred - isInitedBb
-                fgAddRefPred(helperCallBb, isInitedBb);
+    // prevBb always flow into isInitedBb
+    fgAddRefPred(isInitedBb, prevBb);
 
-                // helperCallBb unconditionally jumps to the last block (jumps over fastPathBb)
-                isInitedBb->bbJumpDest = block;
+    // Both fastPathBb and helperCallBb have a single common pred - isInitedBb
+    fgAddRefPred(helperCallBb, isInitedBb);
 
-                //
-                // Re-distribute weights
-                //
+    // helperCallBb unconditionally jumps to the last block (jumps over fastPathBb)
+    isInitedBb->bbJumpDest = block;
 
-                block->inheritWeight(prevBb);
-                isInitedBb->inheritWeight(prevBb);
-                helperCallBb->bbSetRunRarely();
+    //
+    // Re-distribute weights
+    //
 
-                //
-                // Update loop info if loop table is known to be valid
-                //
+    block->inheritWeight(prevBb);
+    isInitedBb->inheritWeight(prevBb);
+    helperCallBb->bbSetRunRarely();
 
-                if (optLoopTableValid && prevBb->bbNatLoopNum != BasicBlock::NOT_IN_LOOP)
-                {
-                    isInitedBb->bbNatLoopNum   = prevBb->bbNatLoopNum;
-                    helperCallBb->bbNatLoopNum = prevBb->bbNatLoopNum;
-                    // Update lpBottom after block split
-                    if (optLoopTable[prevBb->bbNatLoopNum].lpBottom == prevBb)
-                    {
-                        optLoopTable[prevBb->bbNatLoopNum].lpBottom = block;
-                    }
-                }
+    //
+    // Update loop info if loop table is known to be valid
+    //
 
-                // All blocks are expected to be in the same EH region
-                assert(BasicBlock::sameEHRegion(prevBb, block));
-                assert(BasicBlock::sameEHRegion(prevBb, isInitedBb));
+    if (optLoopTableValid && prevBb->bbNatLoopNum != BasicBlock::NOT_IN_LOOP)
+    {
+        isInitedBb->bbNatLoopNum   = prevBb->bbNatLoopNum;
+        helperCallBb->bbNatLoopNum = prevBb->bbNatLoopNum;
+        // Update lpBottom after block split
```

Comment:
Good point, although we still need to update `bbNatLoopNum` for new blocks we created after split.
Also, at this point loop table is invalid so this code is only to make the phase a bit less order-tied 🙂

Reply:
Moved to `fgSplitBlockBeforeTree`

Reply:
I forgot to mention to reuse these methods for `fgExpandRuntimeLookups()` as well.

Reply:
> I forgot to mention to reuse these methods for `fgExpandRuntimeLookups()` as well.

it's done in this PR

Reply:
don't think so...the `fgExpandRuntimeLookups()` seems as it is and using older model of scan_again:

https://github.com/dotnet/runtime/blob/6ca0784d82e49bd2fa8c6607d21f1993e8930b21/src/coreclr/jit/runtimelookup.cpp#L82-L99

I just did that in https://github.com/dotnet/runtime/pull/82973/commits/746fd045b11c5d4133f8b6653f4c81014826ca4b and introduced few generalized methods.

Reply:
then you commented in a wrong thread - this is about updating loop info.

For the IR walk we might want to end up with a generalized way since we already have 4 phases for late call expansion (gc safe points, runtime lookups, static inits, tls)

Reply:
> then you commented in a wrong thread - this is about updating loop info.

Ah yes, I meant it for https://github.com/dotnet/runtime/pull/83911#discussion_r1158791234 and https://github.com/dotnet/runtime/pull/83911#discussion_r1158792731.

Reply:
@kunalspathak btw, don't forget to update your logic to update loop info the same way - you only need to update `bbNatLoopNum` in new blocks now

Reply:
@EgorBo - When I generalized the logic for expansion, I noticed that for runtime lookup you always do the optimization but in others we skip it for rarely run blocks. I think we should skip the expansion of runtime lookup for rarely run blocks too.

Reply:
> @EgorBo - When I generalized the logic for expansion, I noticed that for runtime lookup you always do the optimization but in others we skip it for rarely run blocks. I think we should skip the expansion of runtime lookup for rarely run blocks too.
> 
> Edit: I was hitting an assert in lower, so updated it about the rarely block.

No, it was intentionally enabled everywhere + assert, if you e.g. disable the expansion for no-opts then some test suits will likely timeout, I didn't want to risk and enabled them for rarely executed  blocks too